### PR TITLE
[Yaml] add option to dump objects as maps

### PR DESCRIPTION
**Pull request by xabbuh**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #9870, #12860, #15781, #15937, #16266 |
| License | MIT |
| Doc PR | symfony/symfony-docs#6226 |

### CODE REVIEW

**Comment by xabbuh** [comment]
I am not convinced that we should do this for all objects. Maybe it would be better to only support this option for `stdClass` instances.

**Comment by stof** [comment]
I agree that it should not be done for any object. It makes sense for stdClass, ArrayObject, and maybe other Traversable objects.

DateTime and DateTimeImmutable should be better dumped as a Yaml timestamp (with a parser flag to parse timestamps as a DateTime rather than as an integer too btw)

File: src/Symfony/Component/Yaml/Inline.php [hunk 1]
```diff
@@ -112,6 +112,8 @@ public static function dump($value, $exceptionOnInvalidType = false, $flags = 0)
             case is_object($value):
                 if (Yaml::DUMP_OBJECT & $flags) {
                     return '!php/object:'.serialize($value);
+                } elseif (Yaml::DUMP_OBJECT_AS_MAP & $flags) {
```
Comment (stof) [inline src/Symfony/Component/Yaml/Inline.php]: just ``if`` as the previous if returns

**Comment by xabbuh** [comment]
I updated here to have the mapping only work with `stdClass` and `ArrayObject` instances instead of each and every kind of object.

**Comment by xabbuh** [comment]
see #17836 for the `DateTime` support

**Comment by fabpot** [comment]
This one needs to  be rebased

**Comment by xabbuh** [comment]
rebased, failures are unrelated

**Comment by fabpot** [comment]
Thank you @xabbuh.

**Pull request merged**